### DISCUSSION TITLE
VideoStreamPlayer supported video formats

### DISCUSSION THREAD

**Discussion by RoyalXXX** [godotengine/godot-proposals#9669] [Engine Core]
**VideoStreamPlayer** supports only one format -  **Ogg Theora** (**.ogv**, **VideoStreamTheora**).
But this format is deprecated and does not compress video well. The previous version of the Godot engine supported also **WebM** (**VP8** and **VP9**). **WebM** is a very popular format, why was it removed from the new version? Instead of adding support for new containers (MP4, MKV) and codecs (AV1), you remove what was there before. GDExtension plugins are not an option. 
What's wrong with you? :)

**Comment by sonic2kk** [comment]
From what I recall, video format support in Godot typically comes down to maintainability. I believe there have been discussions on the matter before.

> What's wrong with you? :)

I think the last part is a little uncalled for in what is otherwise a potentially reasonable discussion, if this information is not already well documented.

**Comment by RoyalXXX** [comment]
I really believe that Godot has a chance to compete with proprietary game engines. Just like I used to believe in Blender when others didn't. And I was right. But for this you need to improve the engine, and not remove functions that were in previous versions. Look how many formats Unity supports: https://docs.unity3d.com/Manual/VideoSources-FileCompatibility.html
There are many open source codecs (FFmpeg codecs, AV1, libvpx, x264), why don't you support them?

**Reply by AThousandShips** [reply]
Are all those open source formats? Do they have an open implementation? If not then we simply cannot support them. Nor can we if they are open source but incompatible with the MIT license, like GPL licensed code

**Reply by RoyalXXX** [reply]
If we talk about WebM:
VP8 & VP9 → libvpx (BSD license) 
AV1 → libaom (BSD license)

H.264 → OpenH264 (BSD license)

**Reply by AThousandShips** [reply]
See related:
* https://github.com/godotengine/godot-proposals/issues/7062
* https://github.com/godotengine/godot/pull/53833

**Reply by RoyalXXX** [reply]
> Add MP4/WebM/AV1 import through an FFMPEG conversion wrapper #7062

I have no problem converting any video format to Ogg Theora. I can do it manually as well. So it makes absolutely no sense for a Godot to convert video to Theora using FFmpeg while importing a video. Godot is not a video converter, it's a game engine))

It's important to me that the final game (executable) size is not bloated. Theora is an antiquated format that compresses video very poorly. It's such a bad format that chromium browsers removed support for it due to emerging security risks and because of its extremely low unjustified usage.

**Reply by AThousandShips** [reply]
Then add your voice to that proposal, I just linked relevant proposals... And the reason behind removing the support for WebM

**Reply by Calinou** [reply]
> AV1 → libaom (BSD license)

Note that this reference AV1 decoder/encoder is the slowest one out there.

The fast ones ([dav1d](https://www.videolan.org/projects/dav1d.html)/[rav1e](https://github.com/xiph/rav1e)) are written in Rust, which would be difficult to integrate in Godot's buildsystem (on top of requiring everyone who builds Godot to install a Rust toolchain). The C++ alternative ([SVT-AV1](https://gitlab.com/AOMediaCodec/SVT-AV1/)) contains a lot of inline assembly code, which is hard to maintain from a buildsystem perspective. We've had this experience in the past with WebM which also shipped with a lot of inline assembly. This required users to install a dedicated tool ([yasm](https://yasm.tortall.net/)) to compile Godot, otherwise WebM decoding fell back to C code which was very slow.

That said, AV1 really needs hardware acceleration to be effectively encoded/decoded. Software decoding (or worse, encoding) is really seen as a last resort measure considering how demanding this codec is on the CPU. With SVT-AV1, the highest resolution/framerate that is really viable to be decoded on most CPUs is 1280x720 @ 60 FPS, or 1920x1080 @ 30 FPS. Even then, you'll probably hear your laptop's fans quite a bit in this situation and battery life will suffer on mobile devices.

Hardware-accelerated AV1 decoding is starting to be somewhat widespread, but it's still not universal (especially on Apple hardware) and the APIs are nonstandard. There's Vulkan Video which aims to standardize access to hardware-accelerated video decoding, but it's not gaining as much steam as anticipated and there is no equivalent on other graphics APIs (which Godot supports as of 4.3).

I'd like to see support for modern video formats in core, but I don't see it happening due to the complexity of their implementations.
